File: src/lowlevel/client/icefsoperators/icefsFallocate.cpp
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-30 04:19:29
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-21 02:40:42
 * @FilePath: /icefs/src/lowlevel/client/icefsoperators/icefsFallocate.cpp
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
#include <stdio.h>

#include <string>

#include "icefsClient.hpp"
#include "icefsClientUtils.hpp"

void IcefsClient::DoIcefsFallocate(fuse_req_t fuseReq, fuse_ino_t inode,
                                   int mode, off_t offset, off_t length,
                                   struct fuse_file_info *fi) {
  ICEFS_PR_FUNCTION;

  switch (this->clientConfig.linkType) {
    case ICEFS_LINK_USE_GRPC: {
      icefsgrpc::IcefsFallocateReq req;
      icefsgrpc::IcefsFallocateRes res;
      grpc::ClientContext ctx;

      req.set_mode(mode);
      req.set_offset(offset);
      req.set_length(length);
      req.set_fh(fi->fh);

      grpc::Status status = gRpcClient->DoIcefsFallocate(&ctx, req, &res);
      if (status.ok() && !res.status()) {
        fuse_reply_err(fuseReq, ICEFS_EOK);
      } else {
        fuse_reply_err(fuseReq, res.status() ? res.status() : EIO);
        ICEFS_PR_ERR_STATUS;
      }
      break;
    }

    case ICEFS_LINK_USE_THRIFT: {
      icefsthrift::IcefsFallocateReq req;
      icefsthrift::IcefsFallocateRes res;

      req.__set_mode(mode);
      req.__set_offset(offset);
      req.__set_length(length);
      req.__set_fh(fi->fh);

      icefsThriftConn *thriftConn = thriftClientPool->GetIcefsThriftConn();
      thriftConn->thriftClient->DoIcefsFallocate(res, req);
      thriftClientPool->PutIcefsThriftConn(thriftConn);
      fuse_reply_err(fuseReq, res.status);
      break;
    }

    default:
      fuse_reply_err(fuseReq, EIO);
      break;
  }
}

File: src/lowlevel/server/icefsoperators/icefsFallocate.go
/*
 * @Author: Tan90degrees tangentninetydegrees@gmail.com
 * @Date: 2023-03-11 07:18:32
 * @LastEditors: Tan90degrees tangentninetydegrees@gmail.com
 * @LastEditTime: 2023-04-17 15:55:04
 * @FilePath: /icefs/src/lowlevel/server/icefsoperators/icefsFallocate.go
 * @Description:
 *
 * Copyright (C) 2023 Tan90degrees <tangentninetydegrees@gmail.com>.
 */
package icefsoperators

import (
	"context"
	"icefs-server/icefserror"
	pb "icefs-server/icefsgrpc"
	"icefs-server/icefsthrift"
	"syscall"
)

func (s *IcefsServer) doIcefsFallocate(mode int32, offset int64, length int64, fh uint64) (status int32) {
	return icefserror.IcefsStdErrno(syscall.Fallocate(int(fh), uint32(mode), offset, length))
}

func (s *IcefsGRpcServer) DoIcefsFallocate(ctx context.Context, req *pb.IcefsFallocateReq) (*pb.IcefsFallocateRes, error) {
	var res pb.IcefsFallocateRes

	res.Status = s.server.doIcefsFallocate(req.Mode, req.Offset, req.Length, req.Fh)

	return &res, nil
}

func (s *IcefsThriftServer) DoIcefsFallocate(ctx context.Context, req *icefsthrift.IcefsFallocateReq) (*icefsthrift.IcefsFallocateRes, error) {
	var res icefsthrift.IcefsFallocateRes

	res.Status = s.server.doIcefsFallocate(req.Mode, req.Offset, req.Length, uint64(req.Fh))

	return &res, nil
}
